src/mapping/RadialBasisFctSolver.hpp

```diff
@@ -450,6 +467,113 @@ Eigen::VectorXd RadialBasisFctSolver<RADIAL_BASIS_FUNCTION_T>::solveConsistent(E
+    double      eval              = basisFunction.evaluate(std::sqrt(squaredDifference));
+
+    result += eval * coeffs.row(j.index()).transpose();
+  }
+
+  // 2. we have a separate polynomial, then we have to add it again here;
+  //   out += (_matrixV * polynomialContribution);
+  if (poly.size() > 0) {
+    // constant polynomial
+    result += 1 * poly.row(0).transpose();
+    // the linear contributions
+    int k = 1;
+    for (std::size_t d = 0; d < _localActiveAxis.size(); ++d) {
+      if (_localActiveAxis[d]) {
+        result += out[d] * poly.row(k).transpose();
+        ++k;
+      }
+    }
+  }
+  return result;
+}
+
+template <typename RADIAL_BASIS_FUNCTION_T>
+template <typename IndexContainer>
+void RadialBasisFctSolver<RADIAL_BASIS_FUNCTION_T>::addWriteDataToCache(const mesh::Vertex &v, const Eigen::VectorXd &load, Eigen::MatrixXd &epsilon, Eigen::MatrixXd &Au,
+                                                                        const RADIAL_BASIS_FUNCTION_T &basisFunction, const IndexContainer &inputIDs, const mesh::Mesh &inMesh) const
+{
+  PRECICE_TRACE();
+  // We ignore the dead axis here for the evaluation matrix (matrixA), as the vertex coordinates are zero for a potential 2d case and there is no option in PUM to set them from the user
+
+  // 1. The matrix contribution
+  // Compute RBF values for matrix A
+  PRECICE_ASSERT(Au.rows() == static_cast<Eigen::Index>(inputIDs.size()));
+  PRECICE_ASSERT(Au.cols() == load.size());
+
+  const auto &out = v.rawCoords();
+  for (const auto &j : inputIDs | boost::adaptors::indexed()) {
+    const auto &in                = inMesh.vertex(j.value()).rawCoords();
+    double      squaredDifference = computeSquaredDifference(out, in);
+    double      eval              = basisFunction.evaluate(std::sqrt(squaredDifference));
+
+    Au.row(j.index()) += eval * load.transpose();
+  }
+
+  // 2. we have a separate polynomial, then we have to add it again here;
+  // Eigen::VectorXd epsilon = _matrixV.transpose() * inputData;
+  if (epsilon.size() > 0) {
+    PRECICE_ASSERT(epsilon.rows() == getNumberOfPolynomials());
+    PRECICE_ASSERT(epsilon.cols() == load.size());
+    // constant polynomial
+    epsilon.row(0) += 1 * load.transpose();
+    // the linear contributions
+    int k = 1;
+    for (std::size_t d = 0; d < _localActiveAxis.size(); ++d) {
+      if (_localActiveAxis[d]) {
+        epsilon.row(k) += out[d] * load.transpose();
+        ++k;
+      }
+    }
+  }
+}
+
+template <typename RADIAL_BASIS_FUNCTION_T>
+void RadialBasisFctSolver<RADIAL_BASIS_FUNCTION_T>::evaluateConservativeCache(Eigen::MatrixXd &epsilon, const Eigen::MatrixXd &Au, Eigen::MatrixXd &out) const
+{
+  // mu in the PETSc implementation
+  out = _decMatrixC.solve(Au);
+
+  if (epsilon.size() > 0) {
+    epsilon -= _matrixQ.transpose() * out;
+    // out  = out - solveTranspose tau (sigma in the PETSc impl)
+    out -= static_cast<Eigen::MatrixXd>(_qrMatrixQ.transpose().solve(-epsilon));
```

Comment:
Why the static cast?

Reply:
the result of `solve()` (or in general intermediate results in Eigen) is an Eigen internal type. Sometimes, these things are simply views to delay and merge operations defined on those. In this case, the `-=` operator is not defined on the return type of `solve()`. Hence, we explicitly need to cast (i.e., materialize) the result before using the operator.